### PR TITLE
Group state locale

### PR DESCRIPTION
**Pull request by simonschmeisser**
when parsing the group state `std::stod` is used to convert from `string` to `double` which uses the `c` locale and thus assumes `,` as a decimal separator for some languages.

This PR resurrects the cpp unit tests, ports them to gtest, runs them with classic and Dutch locale and finally removes the call to `std::stod`

In order for the first two commits to fail during unit testing somewhere `sudo locale-gen nl_NL.UTF-8` needs to be added, @rhaschke @davetcoleman you probably know where

### CODE REVIEW

File: test/test_parser.cpp [hunk 1]
```diff
@@ -38,10 +38,7 @@
 #include <urdf_parser/urdf_parser.h>
 #include <fstream>
 #include <stdexcept>
-
-#define EXPECT_TRUE(arg)                                                                                               \
-  if (!(arg))                                                                                                          \
-  throw std::runtime_error("Assertion failed at line " + std::to_string(__LINE__))
+#include <gtest/gtest.h>
```
Comment (rhaschke) [inline test/test_parser.cpp:41]: Thanks for migration to `gtest`!

File: test/srdf_parser_cpp.test [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<launch>
+  <test test-name="test_srdf_parser_cpp" pkg="srdfdom" type="test_cpp">
+    <env name="LC_ALL" value="C" />
+  </test>
+  <test test-name="test_srdf_parser_cpp_locale" pkg="srdfdom" type="test_cpp">
+    <env name="LC_ALL" value="nl_NL.UTF-8" />
```
Comment (rhaschke) [inline test/srdf_parser_cpp.test:6]: > In order for the first two commits to fail during unit testing somewhere `sudo locale-gen nl_NL.UTF-8` needs to be added, @rhaschke @davetcoleman you probably know where

I don't get this comment. Isn't this env setting correctly setting the locale?

**Review by rhaschke — COMMENTED** [review]
Apart from clang-format issues, looks good to me. Thanks for fixing this throughout whole ROS :wink:

File: test/srdf_parser_cpp.test (hunk 2 above)
Comment (simonschmeisser) [reply test/srdf_parser_cpp.test:6]: yes, but in order for the env variable to have an effect the locale needs to be created first

see https://github.com/ros/urdfdom/commit/1857a55897295c1fa46669bf1483dc78cf2467a8 for an example without using moveit-ci

Comment (rhaschke) [reply test/srdf_parser_cpp.test:6]: I see. I added an appropriate command to .travis.yml

**Comment by rhaschke** [comment]
Could you please squash-in the clang-format fix I pushed (I cannot rebase your branch myself). Then this PR should be ready for merging.

**Comment by simonschmeisser** [comment]
I squashed/fixuped the two commits marked fixup

**Comment by rhaschke** [comment]
Thanks. Let's wait for Travis now...

**Comment by rhaschke** [comment]
@simonschmeisser Could you please have another look. One unit test failed.

**Comment by simonschmeisser** [comment]
It works here, I have no idea, sry

**Comment by rhaschke** [comment]
Ok. Needs to postpone this for the weekend then.

**Comment by rhaschke** [comment]
Looks like urdfdom-1.0.0 cannot parse with the locale set. Probably you are already using the newer version of urdfdom locally (as I do).
```
# export LC_ALL=nl_NL.UTF-8; devel/lib/srdfdom/test_cpp 
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from TestCpp
[ RUN      ] TestCpp.testSimple
Error:   Unable to parse component [.001] to a double (while parsing a vector value)
         at line 157 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Could not parse collision element for Link [base_laser_link]
         at line 517 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Unable to parse component [.001] to a double (while parsing a vector value)
         at line 157 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Could not parse collision element for Link [imu_link]
         at line 517 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Unable to parse component [.001] to a double (while parsing a vector value)
         at line 157 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Could not parse collision element for Link [laser_tilt_link]
         at line 517 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/link.cpp
Error:   Unable to parse component [.0244] to a double (while parsing a vector value)
         at line 102 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/pose.cpp
Error:   Malformed parent origin element for joint [r_gripper_led_joint]
         at line 473 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/joint.cpp
Error:   joint xml is not initialized correctly
         at line 206 in /build/urdfdom-ACOcA8/urdfdom-1.0.0/urdf_parser/src/model.cpp
/root/ws_moveit/src/srdfdom/test/test_parser.cpp:74: Failure
Value of: u != NULL
  Actual: false
Expected: true
Segmentation fault (core dumped)
```
I fixed this issue by temporarily setting the C locale for parsing the URDF.

If an unknown locale is set, just a warning is printed, but the test simply passes. There should be at least a warning, e.g. from cmake, that the locale is not available. Otherwise the test is useless.
```
$ export LC_ALL=foo; devel/lib/srdfdom/test_cpp
bash: warning: setlocale: LC_ALL: cannot change locale (foo): No such file or directory
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from TestCpp
[ RUN      ] TestCpp.testSimple
[       OK ] TestCpp.testSimple (13 ms)
[ RUN      ] TestCpp.testComplex
[       OK ] TestCpp.testComplex (6 ms)
[----------] 2 tests from TestCpp (19 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (19 ms total)
[  PASSED  ] 2 tests.
```
@simonschmeisser Can you please extend the cmake file accordingly?

**Comment by simonschmeisser** [comment]
Thanks! Good catch, I'll add a warning on Monday. I should also file a ubuntu bug report to ask for a patch version update now that OSRF finally managed to have releases of urdfdom(_headers) ...

**Comment by simonschmeisser** [comment]
@rhaschke I searched a bit on Monday but did not find a solution for adding this to cmake. If you know a way please feel invited to add it, otherwise I would suggest to merge and add an issue

**Review by davetcoleman — APPROVED** [review]

**Comment by rhaschke** [comment]
I added a simple check for the locale to be available in cmake and squash-merged.

**Comment by simonschmeisser** [comment]
thanks!

**Pull request merged**